New package: darch-0.18.3

pauldotknopf opened the pull request:
This is to use Void Linux with Darch.

https://godarch.com/
https://godarch.com/concepts/recipes/distributions/

This is my first pull request here, so be nice ;)

maxice8 commented:
Can't the GO packages be built using build_style=go ?

pauldotknopf commented:
I tried, but I couldn't get the GOPATH and GOBIN working right.

I tried to refer to existing templates, but they all didn't work. I opted for a less magic, more explicit, approach.

maxice8 commented:
fair

maxice8 commented on srcpkgs/runc/template:
```diff
@@ -0,0 +1,48 @@
+create_wrksrc=yes
+hostmakedepends="git go pkg-config"
+makedepends="libseccomp-devel"
+depends="libseccomp-devel"
```
Is it really necessary to set seccomp-devel to runtime dependencies ? most of the time the package manager will detect that a library is used and add it itself on the proper library.

pauldotknopf replied:
Nope, I don't need to depend on the devel version.

Are you saying that I don't even have to specify a ```depends```? That ```xbs-src``` will ```ldd``` to see the dependencies, and auto establish them for me?

mvf replied:
Exactly. [Documentation Link](https://github.com/voidlinux/void-packages/blob/master/Manual.md#runtime-dependencies)

maxice8 replied:
> Are you saying that I don't even have to specify a depends? That xbs-src will ldd to see the dependencies, and auto establish them for me?

This is the hook in question that does the rdeps generating.

https://github.com/voidlinux/void-packages/blob/master/common/hooks/pre-pkg/04-generate-runtime-deps.sh

It will try it's best but will fail if the program is not cooperative, we recently had to rebuild some packages due to ncurses update from `6.0` -> `6.1`, some packages broke because they didn't have dependencies on the proper ncurses libraries.

maxice8 commented on srcpkgs/containerd/template:
```diff
@@ -0,0 +1,50 @@
+create_wrksrc=yes
+hostmakedepends="git go"
+makedepends="libbtrfs-devel"
+depends="runc libbtrfs"
```
libbrtfs should also be generated from rdeps

maxice8 commented on srcpkgs/runc/template:
```diff
@@ -0,0 +1,48 @@
+do_fetch() {
+	mkdir -p ${wrksrc}/tmp/src/github.com/opencontainers
+	mkdir -p ${wrksrc}/tmp/bin
+	git clone git://github.com/opencontainers/runc ${wrksrc}/tmp/src/github.com/opencontainers/runc
```
I'm not sure this is the best way to fetch the package, i packaged runc pretty quickly and using distfiles= to download the release tarball .xz and it worked (apparently)

pauldotknopf replied:
Yeah. I have been looking at it more, looking into cross compilation.

I think I see my hangup using the ```go``` build style. It also has some cross-build environment variables that needs to be set.

I am going to revamp this templates, will update later today.

Thanks for your help though. I appreciate the time.

pauldotknopf commented:
Guys, I am having troubles with the cross builds. How do I cross compile locally?

```
export XBPS_TARGET_MACHINE=aarch64
export CROSS_BUILD="yes"
./xbps-src pkg runc
```

This doesn't seem to do a cross-compilation.

pauldotknopf commented:
Alright guys, all updated/fixed. Waiting for your review!

maxice8 commented on srcpkgs/containerd/template:
```diff
@@ -0,0 +1,30 @@
+maintainer="Paul Knopf <pauldotknopf@gmail.com>"
+license="Apache-2.0"
+homepage="https://github.com/containerd/containerd"
+distfiles="https://github.com/containerd/containerd/archive/v${_version}.zip"
```
There is a preference for .tar.gz files which don't need the extra unzip dep, if possible change it.

pauldotknopf replied:
No problem.

Making changes to previous commits are difficult, since I can't just do ```git commit --amend```. Since ```containerd``` isn't the tip, can you make sure there isn't anything else that needs to be changed before I make this change?

newbluemoon replied:
You can use `git rebase --interactive` to edit previous commits. See https://help.github.com/articles/using-git-rebase-on-the-command-line/

pauldotknopf commented:
Updated again, ready for review!

Vaelatern commented:
Anywhere with `version=` that isn't `version`, e.g. `subversion`, please make it something like `_subver=` to not confuse certain scripts.

Vaelatern commented:
The `darch` template makes me think it could have been done better.

Vaelatern commented on srcpkgs/darch/template:
```diff
@@ -0,0 +1,40 @@
+	git clone --branch "v${version}" \
+		git://github.com/godarch/darch \
+		${GOSRCPATH}
+}
```
using distfiles

pauldotknopf replied:
to makefile depends on using git.

A new darch version can remove the need, but it is needed now

Vaelatern commented on srcpkgs/darch/template:
```diff
@@ -0,0 +1,40 @@
+	go get github.com/LK4D4/vndr
+	cd ${GOSRCPATH}
+	vndr
+}
```
this sounds like something for the go build style.

pauldotknopf replied:
Add vndr support go he go build style?

pauldotknopf replied:
I forgot to mention.

I had to do it this way because vndr doesn't work with symlink dirs.

Vaelatern commented on srcpkgs/darch/template:
```diff
@@ -0,0 +1,40 @@
+do_build() {
+	cd ${GOSRCPATH}
+	make build TRAVIS_TAG="v${version}" GO_BUILD_FLAGS="-buildmode=pie"
+}
```
Not sure why the makefile is necessary.

pauldotknopf replied:
Building darch requires the use of make.

The built in build stuff for the go build template is too rudimentary.

Vaelatern commented on srcpkgs/darch/template:
```diff
@@ -0,0 +1,40 @@
+do_install() {
+	cd ${GOSRCPATH}
+	make install DESTDIR=${DESTDIR}
+}
```
This I can understand. Except the `cd` at the top with a good build style will not need this.

pauldotknopf replied:
The Makefile exists in the repository, under src/github.com/godarch/darch.

Again, this had to be done because I had to manually setup go path because I can't use distfiles at the moment.

Vaelatern commented on srcpkgs/runc/template:
```diff
@@ -0,0 +1,27 @@
+do_install() {
+	cd ${GOSRCPATH}
+	make install BINDIR=${DESTDIR}/usr/bin
+}
```
I think you want to learn about `build_wrksrc`.
Don't know why you use a `build_style=go` then overwrite it.

pauldotknopf replied:
The Makefile with darch has built in deploy logic.

pauldotknopf replied:
I mainly use the go template to set the GOSRCPATH environment variable.

I can remove build style go and setup GOSRCPATH manually.

Vaelatern replied:
The do build sets up the GOSRCPATH for you as well

pauldotknopf replied:
I tried again removing ```cd ${GOSRCPATH}```, but Go can't find the vendored libraries. The cd has to be there.

pauldotknopf replied:
I am still using build_style=go because it sets up environment variables for cross compilation.

Vaelatern commented on srcpkgs/darch/template:
```diff
@@ -0,0 +1,40 @@
+go_import_path=github.com/godarch/darch
+hostmakedepends="git go"
+depends="containerd"
+only_for_archs="x86_64"
```
This is my biggest complaint.

pauldotknopf replied:
Darch only supports 64 bit.

The executable itself could run cross plat, but to use it, you run docker images. The images only support 64 bit.

Even if someone installs it on arm, it would be useless.

Containers and runc are xplat

Vaelatern replied:
and musl?

pauldotknopf replied:
I can add musl. Forgot about that

Vaelatern replied:
I dislike the use of `only_for_archs` because it doesn't describe the situation at all. At least with a conditional then `broken="reason"` on matching cause, you get a clear explanation that can be checked against upstream's current reality.

Vaelatern replied:
`only_for_archs` seems to say "Yeah, this thing is a binary linked against this exact libc and you will get nowhere if you try to work around this."

pauldotknopf commented:
@Vaelatern, I updated what I could.

* Replaced "verison" with "ver".
* Added musl as a target for Darch. Added a ```broken``` reason for the failure.

Vaelatern commented:
still don't like the use of `only_for_archs`, but it's not so bad it can't be merged. Just so bad that I hope you will maintain it.

pauldotknopf commented:
I use Darch daily. Consider it my hypervisor for Void Linux. I guarantee it will be maintained. :)
